### ISSUE TITLE
An RPM ate my `/bin/*`

### ISSUE THREAD

**Issue by zoidbergwill** [rmohr/bazeldnf#15]
I know using large base images is not very good, and relying on base image binaries as well as RPMs I wanna install is maybe an anti-pattern, but I'm trying to build an image with some RPMs and it ate `/bin/bash` 😭 

What I would've expected, was all the `/bin/*` executables to live in peaceful harmony.

Not sure if we could easily solve this in the rules in a general way?

Before we stumbled upon your Bazel rules, I tried to solve this in a fork of the `rules_docker` scripts around building tars, but it's pretty atrocious:
https://sourcegraph.com/github.com/zoidbergwill/rules_docker@zoid/rpms/-/blob/container/build_tar.py?L289-318

(If we wanted to support something like that, I guess it'd need to be behind an opt-in, so it doesn't break things for people)

Having a minimal reproducible repo would probably be useful, so I'll try strip it down and see if I can reproduce it with some public RPMs sometime.

I hope that all makes sense though!

If I recall correctly, the issue was that `/bin` is a symlink on CentOS 7, so then some of the tars / layering overrides it at some point, or something.

I probably have internal chat history somewhere from when I first tried to fix this a year or so ago. I've managed to avoid it again since then.

**Comment by rmohr** [comment]
Hi, I completely missed your issue request. In general bazeldnf is generating a file system tree which can handle symlinks. However it assumes that it is generating the base layer. Using bazeldnf for in-between layers was not really a focus. Is that what you tried to do?

**Comment by zoidbergwill** [comment]
That is definitely probably the issue, since we're trying to add stuff to an existing image, but generating the image including the base layers instead would probably have some size benefits too, so I'll investigate that.

Thanks!

**Comment by manuelnaranjo** [comment]
Ok I just was bitten by a similar thing, different from @zoidbergwill what I do have is the symlinks into a tar by using pkg_tar from rules_pkg, something like:
```
load("@rules_pkg//pkg:pkg.bzl", "pkg_tar")

pkg_tar(
    name = "symlinks",
    mode = "0755",
    symlinks = {
        "/bin/bazel": "/bin/bazelisk"
    }
)
```

The moment I pass that symlinks into the tars to an image together with an rpmtree my /bin/bash is gone, trying to debug a bit further. I guess given that we both work for the same company, are use cases are similar :D we're installing rpms on top of an already existing base image derived from centos7, and bazeldnf pushes rpm in between layers

**Comment by manuelnaranjo** [comment]
Ohh this explains a LOT:
```
[root@40388c72a4db booking_core_main]# ls -all /bin
lrwxrwxrwx 1 root root 7 Nov 13  2020 /bin -> usr/bin
```

DAMN

**Comment by manuelnaranjo** [comment]
TBH I'm fine with the workaround, not sure how to document it though, it's a tricky case, I hope users will find this issue at least?

**Comment by rmohr** [comment]
@manuelnaranjo could you outline what exactly paths you got? Is this a `pkg_tar` bug? Is it overriding your symlink? I am not sure anymore I understand what you did and what you saw :)

**Comment by rmohr** [comment]
Oh now I think I get it. You had the `pkg_tar` "below" the base image, correct? How about changint the layer order? Would that help?

**Comment by manuelnaranjo** [comment]
> Oh now I think I get it. You had the `pkg_tar` "below" the base image, correct? How about changint the layer order? Would that help?

if you're based on centos7 then this will always happen, the moment you put a symlink into `/bin` you're doomed

**Comment by rmohr** [comment]
if you're based on centos7 then this will always happen, the moment you put a symlink into /bin you're doomed

Hm, this seems like pkg_tar could improve there. Seems like it is not checking if there is a symlink in the tar file.

**Comment by manuelnaranjo** [comment]
You mean on the layer that's being created? pkg_tar has no knowledge of lower layers and the symlink is on a lower layer (base image). Maybe if we would build the first layer with bazeldnf we would be able to not get this issue, but TBH that's not something we can do in the short term, we still have projects outside of bazel that depends on those base images. And we recently started adopting bazeldnf, one day we will but not now.

Having a good understanding of the root cause, and knowing a workaround is good enough for me TBH. BTW I just tested and Ubuntu version I'm running on my workstation has the same type of symlink, so maybe this is a common thing of many distributions?
